text-utils/more.c

```diff
@@ -746,27 +775,123 @@ static cc_t read_user_input(struct more_control *ctl)
 	return c;
 }
 
-/* Read a decimal number from the terminal.  Set cmd to the non-digit
+/* Read a number and command from the terminal.  Set cmd to the non-digit
  * which terminates the number. */
-static int read_number(struct more_control *ctl, cc_t *cmd)
+static struct number_command read_command(struct more_control *ctl)
 {
-	int i;
-	cc_t ch;
-
-	i = 0;
-	ch = ctl->output_tty.c_cc[VKILL];
-	for (;;) {
-		ch = read_user_input(ctl);
-		if (isdigit(ch))
-			i = i * 10 + ch - '0';
-		else if (ch == ctl->output_tty.c_cc[VKILL])
-			i = 0;
-		else {
-			*cmd = ch;
+	cc_t input[8] = { 0 };
+	ssize_t i, ilen;
+	struct number_command cmd = { .key = more_kc_unknown_command };
+
+	if ((ilen = read(STDERR_FILENO, &input, sizeof(input))) <= 0)
+		return cmd;
+	if (2 < ilen) {
+		if (!memcmp(input, "\x1b\x5b\x41\x00", 4)) {		/* arrow up */
+			cmd.key = more_kc_backwards;
+			return cmd;
+		} else if (!memcmp(input, "\x1b\x5b\x42\x00", 4)) {	/* arrow down */
+			cmd.key = more_kc_skip_forward;
+			return cmd;
+		} else if (!memcmp(input, "\x1b\x5b\x35\x7e\x00", 5)) {	/* page up */
+			cmd.key = more_kc_backwards;
+			return cmd;
+		} else if (!memcmp(input, "\x1b\x5b\x36\x7e\x00", 5)) {	/* page down */
+			cmd.key = more_kc_skip_forward;
+			return cmd;
```

Comment:
Not sure, but would it be better to use getch() from ncurses and KEY_NPAGE and KEY_PPAGE (etc)? It seems strange to hardcode it to the code.

Reply:
Another thing I will address in upcoming follow up. That said I will look into using getch().

Reply:
Oh, but to get `getch()` working more(1) would require `initscr()` and that will lead to terminal window. That is not how I imagine more(1). The differentiating factor of more vs less is if terminal window is used or not.

Reply:
Good point.